### ISSUE TITLE
Disabling 'constrained' protection

### ISSUE THREAD

**Issue by Bryantdl7** [FPtje/Falcos-Prop-protection#320]
Is there any documentation laid out of all commands available in FPP? If so point me to that and I should be able to solve this myself if a cvar to disable the constrained repossession of props exists. 

I am trying to hunt down the ability to control contstrained protection, in a sandbox server where people build very complex mechs it would better serve us to just disable the constrained protection which is very valuable in DarkRP, but not as valuable in sandbox where there are little to no world props or other world-owned props at risk of being duped, for example money printer exploits. 

Essentially the problem comes in where they spawn some props without proper prop ownership via E2, and then if they are welded or parented to props they own, they lose ownership of their entire build because it touches something they own. I want to disable this in our specific use case.

**Comment by FPtje** [comment]
Thanks for the request! I don't think there's a setting to disable constrained protection. 

In your post you describe a way to spawn props through e2 that will be assigned to world. That doesn't seem right, shouldn't those props be owned by the owner of the e2?

I feel like if we solve that, you don't need the setting anymore.

**Comment by Bryantdl7** [comment]
After digging deeper with some of our more complex builders it isn't as simple as I first thought. Basically I am able to get the following conclusions: 
1. Certain builds are affected that use complex E2s
2. This can be resolved by using a different prop protection which does not do least privilege constraint protection 
3. The constrained problem happens when the necessary e2 functions to operate these dupes are running

I went as far as to get some example dupes which people lose access to when using FPP, and those are attached. 
[example-dupe-2.txt](https://github.com/FPtje/Falcos-Prop-protection/files/12360318/example-dupe-2.txt)
[example-dupe-1.txt](https://github.com/FPtje/Falcos-Prop-protection/files/12360319/example-dupe-1.txt)

Essentially the tough spot we are stuck in is use a subpar prop protection without constraint protection, or have some complex contraptions unusable. 


Even a simple cvar, or a check box saying "disable constraint protection (not recommended)" would be amazing. Right now until we can get this solved we are using nadmod, and that leaves a lot to be desired... like people being able to physgun world props at any given time.

**Comment by FPtje** [comment]
Thanks for the dupes! I can take a look at it. What is the duplicator addon? Last time I checked there were several.

> Even a simple cvar, or a check box saying "disable constraint protection (not recommended)" would be amazing

I'll take a look at this issue when I have time. If I can't solve the root cause issue quick enough, I'll add the setting 👍

For now I don't think one is much more difficult than the other.

**Comment by Bryantdl7** [comment]
These were both made for advanced duplicator 2.

Sent from my iPhone

On Aug 16, 2023, at 12:56 PM, Falco Peijnenburg ***@***.***> wrote:

﻿

Thanks for the dupes! I can take a look at it. What is the duplicator addon? Last time I checked there were several.

Even a simple cvar, or a check box saying "disable constraint protection (not recommended)" would be amazing

I'll take a look at this issue when I have time. If I can't solve the root cause issue quick enough, I'll add the setting 👍

For now I don't think one is much more difficult than the other.

**Comment by FPtje** [comment]
The dupes paste with some errors (after installing a bunch of addons):

```
[AdvDupe2NotifyPaste]	 Player: (FPtje) Falco Pasted File, example with, 433 Entities and 72 Constraints.
[AdvDupe2Notify]	Pasting...
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
sv: Expression 2 (Sound Interpolation v0.1.0): UDFunction: fade(nnnn) undefined at runtime!
sv: No such function: entity:parentTo(entity) at line 60, char 11
sv: No such function: entity:setAng(angle) at line 90, char 15
sv: No such function: entity:propInertia(vector) at line 6, char 11
sv: No such function: entity:propDrag(number) at line 16, char 11
[AdvDupe2Notify]	Finished Pasting!
Requesting texture value from var "$basetexture" which is not a texture value (material: pp/copy)
```

I couldn't reproduce this with the first dupe, but I could with the second. The cause is that there are some constrained entities owned by world, that should not be taken into account due to them being point entities. This should be fixed for the second dupe in 5d8e657cf643364704ac57c12f006680b2bf98c8. Could you try the latest Github version to see if that works?

**Comment by Bryantdl7** [comment]
It looks like the initial fixes helped with some things, but some other problems have been identified switching back to FPP. I have asked one of our more experienced builders to reach out and open a new ticket for it as the issue is a little more complex though. Hopefully that should be sent in sometime today.

Sent from my iPhone

On Aug 16, 2023, at 4:44 PM, Falco Peijnenburg ***@***.***> wrote:

﻿

The dupes paste with some errors (after installing a bunch of addons):

[AdvDupe2NotifyPaste]    Player: (FPtje) Falco Pasted File, example with, 433 Entities and 72 Constraints.
[AdvDupe2Notify]        Pasting...
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
AD2 PostEntityPaste Error: lua/entities/gmod_wire_expression2/init.lua:595: bad argument #1 to 'pairs' (table expected, got nil)
sv: Expression 2 (Sound Interpolation v0.1.0): UDFunction: fade(nnnn) undefined at runtime!
sv: No such function: entity:parentTo(entity) at line 60, char 11
sv: No such function: entity:setAng(angle) at line 90, char 15
sv: No such function: entity:propInertia(vector) at line 6, char 11
sv: No such function: entity:propDrag(number) at line 16, char 11
[AdvDupe2Notify]        Finished Pasting!
Requesting texture value from var "$basetexture" which is not a texture value (material: pp/copy)


I couldn't reproduce this with the first dupe, but I could with the second. The cause is that there are some constrained entities owned by world, that should not be taken into account due to them being point entities. This should be fixed for the second dupe in 5d8e657<https://github.com/FPtje/Falcos-Prop-protection/commit/5d8e657cf643364704ac57c12f006680b2bf98c8>. Could you try the latest Github version to see if that works?

**Comment by FPtje** [comment]
Closing issue in anticipation of a new issue for the follow up problems
